aws-crt-http/HttpConnection.cs

```diff
@@ -28,39 +30,135 @@ public enum OutgoingBodyStreamState
+        }
+    }
+
+    public class IncomingHeadersDoneEventArgs : HttpClientStreamEventArgs
+    {
+        public bool HasBody { get; private set; }
+
+        internal IncomingHeadersDoneEventArgs(HttpClientStream stream, bool hasBody)
+            : base(stream)
+        {
+            HasBody = hasBody;
+        }
+    }
+
+    public class IncomingBodyEventArgs : HttpClientStreamEventArgs
+    {
+        public byte[] Data { get; private set; }
+        public UInt64 WindowSize { get; private set; }
+
+        internal IncomingBodyEventArgs(HttpClientStream stream, byte[] data, UInt64 windowSize)
+            : base(stream)
+        {
+            Data = data;
+            WindowSize = windowSize;
+        }
+    }
+
+    public class StreamOutgoingBodyEventArgs : HttpClientStreamEventArgs
+    {
+        public byte[] Buffer { get; private set; }
+        public OutgoingBodyStreamState State { get; set; } = OutgoingBodyStreamState.InProgress;
+        public UInt64 BytesWritten { get; set; } = UInt64.MaxValue;
+
+        internal StreamOutgoingBodyEventArgs(HttpClientStream stream, byte[] buffer)
+            : base(stream)
+        {
+            Buffer = buffer;
+        }
+    }
+
+    public class StreamCompleteEventArgs : HttpClientStreamEventArgs
+    {
+        public int ErrorCode { get; private set; }
+
+        internal StreamCompleteEventArgs(HttpClientStream stream, int errorCode)
+            : base(stream)
+        {
+            ErrorCode = errorCode;
+        }
+    }
 
     public sealed class HttpRequestOptions
     {
         public string Method { get; set; }
         public string Uri { get; set; }
         public HttpHeader[] Headers { get; set; }
-        public OnStreamOutgoingBody OnStreamOutgoingBody { get; set; }
-        public OnStreamComplete OnStreamComplete { get; set; }
-        public OnIncomingHeaders OnIncomingHeaders { get; set; }
-        public OnIncomingHeaderBlockDone OnIncomingHeaderBlockDone { get; set; }
-        public OnIncomingBody OnIncomingBody { get; set; }
+        public event EventHandler<StreamOutgoingBodyEventArgs> StreamOutgoingBody;
+        public event EventHandler<StreamCompleteEventArgs> StreamComplete;
+        public event EventHandler<IncomingHeadersEventArgs> IncomingHeaders;
+        public event EventHandler<IncomingHeadersDoneEventArgs> IncomingHeadersDone;
+        public event EventHandler<IncomingBodyEventArgs> IncomingBody;
+
+        internal void Validate()
+        {
+            if (IncomingHeaders == null)
+                throw new ArgumentNullException("IncomingHeaders");
+            if (StreamComplete == null)
+                throw new ArgumentNullException("StreamComplete");
+        }
+
+        internal OutgoingBodyStreamState OnStreamOutgoingBody(HttpClientStream stream, byte[] buffer, out UInt64 bytesWritten)
+        {
+            var e = new StreamOutgoingBodyEventArgs(stream, buffer);
+            StreamOutgoingBody?.Invoke(stream, e);

```

Comment:
Aint C# 4 grand?

Reply:
It's glorious. I spent last night looking at some old C# code of mine, thinking about how much less of it there would be now.